<!DOCTYPE html>
<html lang="en">
<head>
    
    <meta charset="utf-8">
    <title>Home - Documentation</title>
    
    
    <script src="scripts/prettify/prettify.js"></script>
    <script src="scripts/prettify/lang-css.js"></script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc.css">
    <script src="scripts/nav.js" defer></script>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
</head>
<body>

<input type="checkbox" id="nav-trigger" class="nav-trigger" />
<label for="nav-trigger" class="navicon-button x">
  <div class="navicon"></div>
</label>

<label for="nav-trigger" class="overlay"></label>

<nav >
    
    <input type="text" id="nav-search" placeholder="Search" />
    
    <h2><a href="index.html">Home</a></h2><h2><a href="https://github.com/LTLA/bakana.js" target="_blank" class="menu-item" id="repository" >Github repo</a></h2><h3>Classes</h3><ul><li><a href="CellLabellingState.html">CellLabellingState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="CellLabellingState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="CellLabellingState.html#summary">summary</a></li></ul></li><li><a href="ChooseClusteringState.html">ChooseClusteringState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="ChooseClusteringState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="ChooseClusteringState.html#summary">summary</a></li></ul></li><li><a href="CustomSelectionsState.html">CustomSelectionsState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="CustomSelectionsState.html#addSelection">addSelection</a></li><li data-type='method' style='display: none;'><a href="CustomSelectionsState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="CustomSelectionsState.html#fetchResults">fetchResults</a></li><li data-type='method' style='display: none;'><a href="CustomSelectionsState.html#removeSelection">removeSelection</a></li><li data-type='method' style='display: none;'><a href="CustomSelectionsState.html#summary">summary</a></li></ul></li><li><a href="FeatureSelectionState.html">FeatureSelectionState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="FeatureSelectionState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="FeatureSelectionState.html#summary">summary</a></li></ul></li><li><a href="InputsState.html">InputsState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="InputsState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="InputsState.html#fetchAnnotations">fetchAnnotations</a></li><li data-type='method' style='display: none;'><a href="InputsState.html#summary">summary</a></li></ul></li><li><a href="KmeansClusterState.html">KmeansClusterState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="KmeansClusterState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="KmeansClusterState.html#summary">summary</a></li></ul></li><li><a href="MarkerDetectionState.html">MarkerDetectionState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="MarkerDetectionState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="MarkerDetectionState.html#fetchGroupResults">fetchGroupResults</a></li><li data-type='method' style='display: none;'><a href="MarkerDetectionState.html#numberOfGroups">numberOfGroups</a></li><li data-type='method' style='display: none;'><a href="MarkerDetectionState.html#summary">summary</a></li></ul></li><li><a href="NeighborIndexState.html">NeighborIndexState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="NeighborIndexState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="NeighborIndexState.html#summary">summary</a></li></ul></li><li><a href="NormalizationState.html">NormalizationState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="NormalizationState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="NormalizationState.html#fetchExpression">fetchExpression</a></li><li data-type='method' style='display: none;'><a href="NormalizationState.html#summary">summary</a></li></ul></li><li><a href="PcaState.html">PcaState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="PcaState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="PcaState.html#summary">summary</a></li></ul></li><li><a href="QualityControlState.html">QualityControlState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="QualityControlState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="QualityControlState.html#fetchFilteredAnnotations">fetchFilteredAnnotations</a></li><li data-type='method' style='display: none;'><a href="QualityControlState.html#fetchFilteredBlock">fetchFilteredBlock</a></li><li data-type='method' style='display: none;'><a href="QualityControlState.html#summary">summary</a></li></ul></li><li><a href="SnnGraphClusterState.html">SnnGraphClusterState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="SnnGraphClusterState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="SnnGraphClusterState.html#summary">summary</a></li></ul></li><li><a href="TsneState.html">TsneState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="TsneState.html#animate">animate</a></li><li data-type='method' style='display: none;'><a href="TsneState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="TsneState.html#summary">summary</a></li></ul></li><li><a href="UmapState.html">UmapState</a><ul class='methods'><li data-type='method' style='display: none;'><a href="UmapState.html#animate">animate</a></li><li data-type='method' style='display: none;'><a href="UmapState.html#compute">compute</a></li><li data-type='method' style='display: none;'><a href="UmapState.html#summary">summary</a></li></ul></li></ul><h3>Global</h3><ul><li><a href="global.html#analysisDefaults">analysisDefaults</a></li><li><a href="global.html#availableReaders">availableReaders</a></li><li><a href="global.html#callScran">callScran</a></li><li><a href="global.html#createAnalysis">createAnalysis</a></li><li><a href="global.html#createKanaFile">createKanaFile</a></li><li><a href="global.html#freeAnalysis">freeAnalysis</a></li><li><a href="global.html#initialize">initialize</a></li><li><a href="global.html#loadAnalysis">loadAnalysis</a></li><li><a href="global.html#parseKanaFile">parseKanaFile</a></li><li><a href="global.html#removeHDF5File">removeHDF5File</a></li><li><a href="global.html#runAnalysis">runAnalysis</a></li><li><a href="global.html#saveAnalysis">saveAnalysis</a></li><li><a href="global.html#setCellLabellingDownload">setCellLabellingDownload</a></li><li><a href="global.html#setCreateLink">setCreateLink</a></li><li><a href="global.html#setResolveLink">setResolveLink</a></li><li><a href="global.html#setVisualizationAnimate">setVisualizationAnimate</a></li><li><a href="global.html#terminate">terminate</a></li><li><a href="global.html#validateAnnotations">validateAnnotations</a></li></ul>
</nav>

<div id="main">
    

    



    


    <section class="package">
        <h3> </h3>		
    </section>









    



    <section class="readme usertext">
        <article><h1>Backend for kana</h1>
<h2>Overview</h2>
<p><strong>bakana</strong> provides the compute backend for the <a href="https://github.com/jkanche/kana"><strong>kana</strong></a> application.
It provides a pipeline for a routine single-cell RNA-seq analysis, starting from the count matrix and finishing with the usual results (markers, clusters, t-SNE and so on).
Datasets involving multiple samples in one or multiple matrices can also be analyzed with blocking and batch correction.
The pipeline can be executed both in the browser and on Node.js.
It supports in-memory caching of the analysis state for fast iterative re-analysis,
as well as serialization of the state for storage and distribution to other machines.</p>
<h2>Getting started</h2>
<p>Install the package from <a href="https://npmjs.com/package/bakana">NPM</a> using the usual method:</p>
<pre class="prettyprint source lang-sh"><code>npm install bakana
</code></pre>
<p>See the <a href="https://ltla.github.io/bakana/">reference documentation</a> for details on available functions.</p>
<h2>Running analyses</h2>
<p>We perform an analysis with the following commands:</p>
<pre class="prettyprint source lang-js"><code>import * as bakana from &quot;bakana&quot;;
await bakana.initialize({ numberOfThreads: 8 });

let state = await bakana.createAnalysis();
let params = bakana.analysisDefaults();

await bakana.runAnalysis(state, 
    // Specify files using paths (Node.js) or File objects (browser).
    { my_data: { format: &quot;10X&quot;, h5: &quot;/some/file/path&quot; } },
    params
);
</code></pre>
<p>Each step is represented by a <code>*State</code> class instance as a property of <code>state</code>, containing the analysis results.
We can extract summaries for diagnostics or display on a UI:</p>
<pre class="prettyprint source lang-js"><code>state.quality_control.summary();
// {
//   data: {
//     default: {
//       sums: [Float64Array],
//       detected: [Int32Array],
//       proportion: [Float64Array]
//     }
//   },
//   ...
// }

state.pca.summary():
// {
//   var_exp: Float64Array(20) [...]
// }

// Some summary() are promises, see the relevant documentation.
await state.tsne.summary();
// {
//   x: [Float64Array],
//   y: [Float64Array],
//   iterations: 500
// }
</code></pre>
<p>Alternatively, we can supply a callback that processes summaries from each step as soon as it finishes.
This can be useful for, e.g., posting diagnostics to another system once they become available.</p>
<pre class="prettyprint source lang-js"><code>function finisher(step, results) {
    console.log(step);
    // Do other stuff with 'results' here.
}

await bakana.runAnalysis(state, 
    // Specify files using paths (Node.js) or File objects (browser).
    { my_data: { format: &quot;10X&quot;, h5: &quot;/some/file/path&quot; } },
    params,
    { finishFun: finisher }
);
// inputs
// quality_control
// normalizaton
// feature_selection
// pca
// neighbor_index
// kmeans_cluster
// ...
</code></pre>
<p>If the analysis is re-run with different parameters, <strong>bakana</strong> will only re-run the affected steps.
This includes all steps downstream of any step with changed parameters.</p>
<pre class="prettyprint source lang-js"><code>params.pca.num_pcs = 15;

await bakana.runAnalysis(state, 
    // Specify files using paths (Node.js) or File objects (browser).
    { my_data: { format: &quot;10X&quot;, h5: &quot;/some/file/path&quot; } },
    params,
    { finishFun: finisher }
);
// pca
// neighbor_index
// ...
</code></pre>
<h2>Saving analyses</h2>
<p>Given an analysis state, we can save the parameters and results to a HDF5 file.</p>
<pre class="prettyprint source lang-js"><code>let collected = await bakana.saveAnalysis(state, &quot;whee.h5&quot;);
</code></pre>
<p><code>saveAnalysis</code> will return a promise that resolves to an array of paths (Node.js) or <code>File</code> objects for the original data files.
These can be used to assemble a <code>*.kana</code>-format file following the <a href="https://ltla.github.io/kanaval"><strong>kanaval</strong> specification</a>.
By default, this assumes that we are embedding the data files into the <code>*.kana</code> file.</p>
<pre class="prettyprint source lang-js"><code>let res = await bakana.createKanaFile(&quot;whee.h5&quot;, collected.collected);
</code></pre>
<p>Advanced users may prefer to store links to the data files rather than embedding them.
This can be achieved using the <code>setCreateLink()</code> function to define a mechanism for creating application-specific links.
For example, the <strong>kana</strong> application uses IndexedDB to cache each file for later use in the browser.</p>
<pre class="prettyprint source lang-js"><code>bakana.setCreateLink(save_to_some_db);
bakana.setResolveLink(load_from_some_db);
</code></pre>
<h2>Loading analyses</h2>
<p>Given a <code>*.kana</code> file, it is straightforward to extract the various state and data files.</p>
<pre class="prettyprint source lang-js"><code>let loader = await bakana.parseKanaFile(&quot;something.kana&quot;, &quot;foo.h5&quot;);
</code></pre>
<p>Once the <code>*.kana</code> file is parsed, we can reload the analysis state into memory.
This will also report the parameters used at each step.</p>
<pre class="prettyprint source lang-js"><code>let reloaded = await bakana.loadAnalysis(&quot;foo.h5&quot;, loader);
let new_state = reloaded.state;
let new_params = reloaded.parameters;
</code></pre>
<p>These can be used to perform a new analysis, possibly after modifying some parameters.</p>
<pre class="prettyprint source lang-js"><code>new_params.pca.num_hvgs = 3000;
await bakana.runAnalysis(new_state, null, new_params);
</code></pre>
<p>Again, we can supply a callback that runs on the results of each step as soon as they are loaded.</p>
<pre class="prettyprint source lang-js"><code>let reloaded = await bakana.loadAnalysis(&quot;whee.h5&quot;, loader, { finishFun: finisher });
</code></pre>
<p>If links are present, it is assumed that the application will use <code>setResolveLink()</code> to specify a mechanism to resolve each link to a data file.</p>
<h2>Terminating analyses</h2>
<p>Once a particular analysis is finished, we should free the resources of its state.</p>
<pre class="prettyprint source lang-js"><code>bakana.freeAnalysis(state);
</code></pre>
<p>If all analyses are complete, we can terminate the entire session.
This is necessary on Node.js to allow the runtime to exit properly.</p>
<pre class="prettyprint source lang-js"><code>bakana.terminate();
</code></pre>
<h2>Adding custom readers</h2>
<p>Developers can add new data readers to supplement the defaults in <code>src/readers</code>.
This is most useful for pulling data from external databases for entry into the <strong>bakana</strong> workflow.
A new reader should be implemented as an ES6 module with the required exports below.</p>
<p><strong><code>abbreviate(args)</code>:</strong> generate a summary of the matrix to check whether the inputs have changed on re-analysis.</p>
<ul>
<li>This should accept an object <code>args</code> containing information about a single count matrix in the reader's desired format.
<code>args</code> may contain an arbitrary number of properties - their specification is left to the reader.
By convention, properties referring to files should be <code>File</code> objects in the browser and strings containing file paths for Node.js.</li>
<li>This should quickly return an object that summarizes the information about the count matrix.
The returned object should be easily stringified for quick comparison,
and should be unique enough to distinguish between most other values of <code>args</code>.
We generally recommend generating a summary based on the file name and size, which is fast and unique enough for most comparisons.</li>
</ul>
<p><strong><code>preflight(args)</code>:</strong> fetch annotations from the matrix before the analysis begins.</p>
<ul>
<li>This should accept an object <code>args</code> containing information about a single count matrix in the reader's desired format.
The expected structure of <code>args</code> is identical to that described for <code>abbreviate()</code>.</li>
<li>This should return an object containing the <code>genes</code> and <code>annotations</code> properties.
<code>genes</code> should be an object where each key is a gene annotation field name and each value is an array of per-gene information, usually strings containing Ensembl identifiers or symbols.
<code>annotations</code> should be an array of strings containing the names of the per-cell annotation fields.</li>
<li>Alternatively, this method may return a promise that resolves to such an object.</li>
</ul>
<p><strong><code>Reader(args)</code>:</strong> the <code>Reader</code> class, where each instance contains all information about a single count matrix.</p>
<ul>
<li>This constructor should accept an object <code>args</code> containing information about a single count matrix in the reader's desired format.
The expected structure of <code>args</code> is identical to that described for <code>abbreviate()</code>.</li>
<li>The structure of the class is left as an implementation detail for each reader.</li>
</ul>
<p><strong><code>Reader.format()</code>:</strong> specify the format of the count matrix.</p>
<ul>
<li>This should return a string specifying the format of the count matrix corresponding to this reader.
The value of the string is defined by the reader's developer.</li>
</ul>
<p><strong><code>Reader.load()</code>:</strong> load the count matrix into memory.</p>
<ul>
<li>This should return an object containing <code>matrix</code>, a <code>ScranMatrix</code> object.
The object may also contain <code>genes</code>, an object where each key is a gene annotation field name and each value is an array of per-gene information;
and/or <code>annotations</code>, an object where each key is a cell annotation field name and each value is an array of per-cell information.</li>
<li>Alternatively, this method may return a promise that resolves to such an object.</li>
</ul>
<p><strong><code>Reader.serialize(embeddedSaver)</code>:</strong> register the count matrix in the state file.</p>
<ul>
<li>This should return an array of objects where each object represents a data file used to create the count matrix.
Each object should contain <code>type</code>, a string specifying the type of file; and <code>name</code>, the name of the file.
Alternatively, this method may return a promise that resolves to this array.</li>
<li>If <code>embeddedSaver</code> is a function, it should be called on each file, in the same order as they occur in the returned array.
<code>embeddedSaver</code> will accept two arguments - a file and its size - and will return a Promise that resolves to an object containing <code>offset</code> and <code>size</code>.
(Here, a &quot;file&quot; is defined as an <code>ArrayBuffer</code> containing the contents of the file in the browser, or a path to a file in Node.js.)
The reported <code>offset</code> and <code>size</code> should then be included in the properties of the corresponding object in the returned array.</li>
<li>If <code>embeddedSaver = null</code>, each object in the returned array should instead contain <code>id</code>.
This should be a unique string, typically containing a link to some external database system that holds the corresponding file.
The interpretation of <code>id</code> is left to the reader, as long as it can be used to meaningfully recover the same file in <code>unserialize()</code> below.</li>
</ul>
<p><strong><code>unserialize(values, embeddedLoader)</code>:</strong> load the count matrix from the state file.</p>
<ul>
<li><code>values</code> should be an array of objects where object represents a data file used to create the count matrix.
Each object should contain <code>type</code> and <code>name</code>, as described in the <code>Reader.serialize()</code> method.</li>
<li>If <code>embeddedLoader</code> is a function, each object in <code>values</code> will additionally contain <code>offset</code> and <code>size</code> integers.
<code>embeddedLoader</code> will accept two arguments - the offset and size - and will return the corresponding file.
(Again, a &quot;file&quot; is defined as an <code>ArrayBuffer</code> containing the contents of the file in the browser, or a path to a file in Node.js.)
<code>embeddedLoader</code> should be called on each entry of <code>values</code> in order - note, this should be done even if not all files are used.</li>
<li>If <code>embeddedLoader = null</code>, each object in <code>values</code> will additionally contain an <code>id</code> string.
This usually contains a link to some external database system as defined by <code>Reader.serialize()</code>.</li>
<li>This method should return an instance of the <code>Reader</code> class containing all information related to the count matrix represented by <code>values</code>.
Presumably this is done using the file contents returned by <code>embeddedLoader</code> or the linked files from the <code>id</code> values.
Alternatively, this method may return a promise that resolves to a <code>Reader</code> instance.</li>
</ul>
<p>Applications can then register custom readers by assigning to the <code>bakana.availableReaders</code> object.
The property name should be the name of the reader format (the same as that returned by <code>Reader.format()</code>) and the value should be the ES6 module object implementing the reader.</p>
<h2>Developer notes</h2>
<p>Testing requires some combination of the options below,
depending on the version of Node.js available.</p>
<pre class="prettyprint source lang-sh"><code>node --experimental-vm-modules \
    --experimental-wasm-threads \
    --experimental-wasm-bulk-memory \
    --experimental-wasm-bigint \
    node_modules/jest/bin/jest.js \
    --testTimeout=100000000 \
    --runInBand
</code></pre></article>
    </section>






    
    
</div>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.6.10</a> on Sat Apr 09 2022 23:56:33 GMT+0000 (Coordinated Universal Time) using the <a href="https://github.com/clenemt/docdash">docdash</a> theme.
</footer>

<script>prettyPrint();</script>
<script src="scripts/polyfill.js"></script>
<script src="scripts/linenumber.js"></script>

<script src="scripts/search.js" defer></script>


<script src="scripts/collapse.js" defer></script>


</body>
</html>